pkg/apis/certmanager/v1/types_certificate.go

```diff
@@ -156,6 +156,17 @@ type CertificateSpec struct {
 	// +optional
 	IsCA bool `json:"isCA,omitempty"`
 
+	// MaxPathLen sets the pathLen in a certificate's basicConstraints, controlling the maximum
+	// length of the chain of certificates which can be issued by this certificate.
+	// If a certificate's pathLen is 0, then it cannot issue CA certificates; if pathLen is n > 0,
+	// this certificate may issue a CA certificate, which in turn may issue a CA certificate only if
+	// n-1 is greater than zero.
+	// Self-signed root certificates will generally not have a MaxPathLen, while most intermediates
+	// will usually want to have a pathLen set to the smallest value possible.
+	// If omitted, no pathLen will be added to the issued certificate.
```

Comment:
The language here gives the impression that pathlen will always be set by the issuer. Can this be update to state that pathlen will be _requested where possible_, and not all issuers support it.

Currently we are only supporting the SelfSigned and CA issuer. Can this not be extended to others?

Reply:
Why is this field called `MaxPathLen` and not `PathLen` when we are just copying this value over to the template anyway? This will also causes semantic issues for other issuers if they support it since they will likely just accept a `PathLen` value, not a "`MaxPathLen`".

Reply:
>  Why is this field called MaxPathLen and not PathLen when we are just copying this value over to the template anyway?

Mostly because that's what it's called in the template:

```go
go doc x509 Certificate | grep MaxPathLen
        MaxPathLen int
        MaxPathLenZero bool
```

I don't feel strongly about the name either way; the RFCs talk about "path lengths", openssl renders as pathLen, and Go talks about MaxPathLen.

I think `MaxPathLen` is probably more descriptive of what the thing actually is. "Max" hints that we're referring to a flexible limit here; `PathLen=1` might reasonably be interpreted as being a requirement  that there be exactly 1 CA in the path, whereas it actually means there can be up to 1 CA in the path.

> This will also causes semantic issues for other issuers if they support it since they will likely just accept a PathLen value, not a "MaxPathLen".

This is _really_ quite far out there as far as concerns go.

Vault wants `max_path_length`: https://www.vaultproject.io/api-docs/secret/pki#generate-root

I haven't checked Venafi or any others yet

Reply:
> Currently we are only supporting the SelfSigned and CA issuer. Can this not be extended to others?

Vault can do it; this is another reason the PR is a draft

Reply:
Makes sense!

Should we be doing more to ensure that the pathlen of a signed cert is less than the issuer? Is this already covered in Go?

Reply:
> Should we be doing more to ensure that the pathlen of a signed cert is less than the issuer? Is this already covered in Go?

That would certainly be a nice addition, yeah. I'll look into it. Go doesn't seem to be very strict on it on the issuance side from a quick glance, although there are some TODO comments talking about it:

```console
$ go doc -src x509 Certificate.CheckSignatureFrom 
package x509 // import "crypto/x509"

// CheckSignatureFrom verifies that the signature on c is a valid signature
// from parent.
func (c *Certificate) CheckSignatureFrom(parent *Certificate) error {
...
	// TODO(agl): don't ignore the path length constraint.
...
}
```